### ISSUE TITLE
`[[gnu::warn_unused]]` not considered in `-Wunused-private-field`

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#62472]
```cpp
class [[gnu::warn_unused]] S
{
public:
    S();
};

void f()
{
    int i = 0;
    S s;
}

class C
{
private:
    const int i = 0;
    const S s;
};
```

```console
<source>:9:9: warning: unused variable 'i' [-Wunused-variable]
    int i = 0;
        ^
<source>:10:7: warning: unused variable 's' [-Wunused-variable]
    S s;
      ^
<source>:16:15: warning: private field 'i' is not used [-Wunused-private-field]
    const int i = 0;
              ^
```
https://godbolt.org/z/nqeqMYq5E

**Comment by a-tarasyuk** [comment]
@firewave should a warning be issued for `s` or skipped for `i`? Is there any documentation or spec that provides more details on the expected behavior before proceeding with the PR?

```cpp
class C
{
private:
    const int i = 0;
    const S s;
};
```

/cc @erichkeane

**Comment by firewave** [comment]
> [@firewave](https://github.com/firewave) should a warning be issued for `s` or skipped for `i`?

This is about `C::s` not triggering a warning.

> Is there any documentation or spec that provides more details on the expected behavior before proceeding with the PR?

There is currently "no documentation" - see https://clang.llvm.org/docs/AttributeReference.html#warn-unused.

**Comment by erichkeane** [comment]
It seems to me that `s` should be warned on, or neither should (that is, both or neither).  BUT I see the https://github.com/llvm/llvm-project/commit/a4fbc091846206fb6f5bc36115d65075764b51ea was about `maybe_unused` (which I apparently missed before break when I took a look at this).  `maybe_unused` should still suppress it, and `warn_unused` should likely warn all the time, and a 'naked' type should warn unless it has side-effects (that is, a non-trivial ctor/dtor).
